io/private/EptSupport.hpp

```diff
@@ -90,6 +158,127 @@ inline std::array<double, 3> toArray3(const Json::Value& json)
     return result;
 }
 
+class PDAL_DLL Key
+{
+    // An EPT key representation (see https://git.io/fAiBh).  A depth/X/Y/Z key
+    // representing a data node, as well as the bounds of the contained data.
+public:
+    Key(BOX3D b) : b(b) { }
+    Key(std::string s)
+    {
+        const std::vector<std::string> tokens(Utils::split(s, '-'));
+        if (tokens.size() != 4)
+            throw std::runtime_error("Invalid EPT KEY: " + s);
+        d = std::stoull(tokens[0]);
+        x = std::stoull(tokens[1]);
+        y = std::stoull(tokens[2]);
+        z = std::stoull(tokens[3]);
+    }
+
+    BOX3D b;
+    uint64_t d = 0;
+    uint64_t x = 0;
+    uint64_t y = 0;
+    uint64_t z = 0;
+
+    std::string toString() const
+    {
+        return std::to_string(d) + '-' + std::to_string(x) + '-' +
+            std::to_string(y) + '-' + std::to_string(z);
+    }
+
+    double& operator[](uint64_t i)
+    {
+        switch (i)
+        {
+            case 0: return b.minx;
+            case 1: return b.miny;
+            case 2: return b.minz;
+            case 3: return b.maxx;
+            case 4: return b.maxy;
+            case 5: return b.maxz;
+            default: throw pdal_error("Invalid Key[] index");
+        }
+    }
+
+    uint64_t& idAt(uint64_t i)
+    {
+        switch (i)
+        {
+            case 0: return x;
+            case 1: return y;
+            case 2: return z;
+            default: throw pdal_error("Invalid Key::idAt index");
+        }
+    }
+
+    Key bisect(uint64_t direction) const
+    {
+        Key key(*this);
+        ++key.d;
+
+        auto step([&key, direction](uint8_t i)
+        {
+            key.idAt(i) *= 2;
+
+            const double mid(key[i] + (key[i + 3] - key[i]) / 2.0);
+            const bool positive(direction & (((uint64_t)1) << i));
+            if (positive)
+            {
+                key[i] = mid;
+                ++key.idAt(i);
+            }
+            else
+            {
+                key[i + 3] = mid;
+            }
+        });
+
+        for (uint8_t i(0); i < 3; ++i)
+            step(i);
+
+        return key;
+    }
+};
+
+using EptHierarchy = std::map<Key, uint64_t>;
+
+class PDAL_DLL Addon
+{
+public:
+    Addon(const PointLayout& layout, const arbiter::Endpoint& ep,
+            Dimension::Id id)
+        : m_ep(ep)
+        , m_id(id)
+        , m_type(layout.dimType(m_id))
+        , m_size(layout.dimSize(m_id))
+        , m_name(layout.dimName(m_id))
+    { }
+
+    const arbiter::Endpoint& ep() const { return m_ep; }
+    Dimension::Id id() const { return m_id; }
+    Dimension::Type type() const { return m_type; }
+    uint64_t size() const { return m_size; }
+    const std::string& name() const { return m_name; }
+
+    EptHierarchy& hierarchy() { return m_hierarchy; }
+    uint64_t count(const Key& key) const
+    {
+        if (m_hierarchy.count(key)) return m_hierarchy.at(key);
```

Comment:
IMO this is harder to read than if you put the `return` on the next line.

Reply:
And why can't you just return `m_hierarchy.count(key)` ?

Reply:
`m_hierarchy.count(key)` would return 1 or 0 - `Addon::count` needs to return the number of points for that key.  I'll rename `Addon::count` to `Addon::points`, which would be less confusing given the internal `std::map` usage.

Reply:
Fixed at fce635d.